tr1/ui: add ingame settings

rr- opened the pull request:
#### Checklist

- [x] I have read the [coding conventions](https://github.com/LostArtefacts/TRX/blob/master/CONTRIBUTING.md#coding-conventions)
- [x] I have added a changelog entry about what my pull request accomplishes, or it is an internal change
- [x] I have added a readme entry about my new feature or OG bug fix, or it is a different change

#### Description

This is just a draft to bring every config tool option into ingame settings.
The options look ridiculous at the moment and I'd appreciate suggestions on rewording the labels (including enum values) to keep the dialogs sleek. As it is, in my opinion this PR is quite a downgrade to the dialogs – compare TR2 vs TR1:

![20250601_192804_Level_0](https://github.com/user-attachments/assets/37c20fed-8f78-4651-ba9e-65692bc1a88c)

![20250601_192813_Level_0](https://github.com/user-attachments/assets/7ef1990f-2e92-4969-aa0d-821088d84b3a)


I'm more than open to move some settings around, too.
Changing some settings may lead to crashes – we need to settle these too prior to merging this.

TR2 not touched just yet.

rr- committed:
tr1/ui: add ingame settings

rr- commented:
It doesn't help that the "scale" of the settings is very different – enabling Lara's braid or wind is a major change, whereas crystals PS1 tint… feels really inadequate.

aredfan commented:
- [x] `Healthbar mode` incorrectly includes a `Boss only` option.
- [x] `Airbar mode` has more options than expected.
- [x] `Enemy bar mode` has more options than expected.

rr- commented:
> * [ ]  `Healthbar mode` incorrectly includes a `Boss only` option.
> * [ ]  `Airbar mode` has more options than expected.
> * [ ]  `Enemy bar mode` has more options than expected.

@aredfan thanks – fixed

aredfan commented on src/tr1/game/game_string.def:
```diff
@@ -23,6 +15,146 @@ GS_DEFINE(OSD_TEXTURE_FILTER_SET, "Texture filter set to %s")
+GS_DEFINE(GAMEPLAY_SETTINGS_START_LARA_HITPOINTS,           "Lara's starting health")
+GS_DEFINE(GAMEPLAY_SETTINGS_DISABLE_MEDPACKS,               "Remove medipacks")
+GS_DEFINE(GAMEPLAY_SETTINGS_DISABLE_HEALING_BETWEEN_LEVELS, "Do not heal on level finish")
+GS_DEFINE(GAMEPLAY_SETTINGS_DISABLE_SHOTGUN,                "Remove shotgun")
+GS_DEFINE(GAMEPLAY_SETTINGS_DISABLE_MAGNUMS,                "Remove magnums")
+GS_DEFINE(GAMEPLAY_SETTINGS_DISABLE_UZIS,                   "Remove Uzis")
```
Just a small nitpick here, it'll be worth making `Uzis` all lower case to be consistent with the `shotgun` and `magnums` options.

lahm86 commented on src/tr1/game/game_string.def:
```diff
@@ -23,6 +15,146 @@ GS_DEFINE(OSD_TEXTURE_FILTER_SET, "Texture filter set to %s")
+GS_DEFINE(GAMEPLAY_SETTINGS_ENABLE_TIMER_IN_INVENTORY,      "Timer counts in inventory")
+GS_DEFINE(GAMEPLAY_SETTINGS_ENABLE_ITEM_EXAMINING,          "Item examination")
+GS_DEFINE(GAMEPLAY_SETTINGS_ENABLE_TR3_SIDESTEPS,           "Enhanced sidesteps")
+GS_DEFINE(GAMEPLAY_SETTINGS_ENABLE_RESPONSIVE_PASSPORT,     "Enable more responsive passport")
```
`Responsive passport`

lahm86 commented on src/tr1/game/game_string.def:
```diff
@@ -23,6 +15,146 @@ GS_DEFINE(OSD_TEXTURE_FILTER_SET, "Texture filter set to %s")
+GS_DEFINE(GAMEPLAY_SETTINGS_ENABLE_TARGET_CHANGE,           "Target change")
+GS_DEFINE(GAMEPLAY_SETTINGS_TARGET_MODE,                    "Weapon target lock mode")
+GS_DEFINE(GAMEPLAY_SETTINGS_ENABLE_LEAN_JUMPING,            "Lean jumping")
+GS_DEFINE(GAMEPLAY_SETTINGS_ENABLE_NUMERIC_KEYS,            "Numeric key quick item use")
```
`Hotkey item use`

lahm86 commented on src/tr1/game/game_string.def:
```diff
@@ -23,6 +15,146 @@ GS_DEFINE(OSD_TEXTURE_FILTER_SET, "Texture filter set to %s")
+GS_DEFINE(GAMEPLAY_SETTINGS_ENABLE_TR2_SWIM_CANCEL,         "Responsive swim cancel")
+GS_DEFINE(GAMEPLAY_SETTINGS_ENABLE_WADING,                  "Wading")
+GS_DEFINE(GAMEPLAY_SETTINGS_ENABLE_TARGET_CHANGE,           "Target change")
+GS_DEFINE(GAMEPLAY_SETTINGS_TARGET_MODE,                    "Weapon target lock mode")
```
`Weapon lock mode`

lahm86 commented on src/tr1/game/game_string.def:
```diff
@@ -23,6 +15,146 @@ GS_DEFINE(OSD_TEXTURE_FILTER_SET, "Texture filter set to %s")
+GS_DEFINE(GAMEPLAY_SETTINGS_ENABLE_CONSOLE,                 "Console")
+GS_DEFINE(GAMEPLAY_SETTINGS_START_LARA_HITPOINTS,           "Lara's starting health")
+GS_DEFINE(GAMEPLAY_SETTINGS_DISABLE_MEDPACKS,               "Remove medipacks")
+GS_DEFINE(GAMEPLAY_SETTINGS_DISABLE_HEALING_BETWEEN_LEVELS, "Do not heal on level finish")
```
- `Enable perma-damage`
- `Persistent damage`
- `Carry forward HP`

rr- replied:
Went with `Persistent damage`

lahm86 commented on src/tr1/game/game_string.def:
```diff
@@ -23,6 +15,146 @@ GS_DEFINE(OSD_TEXTURE_FILTER_SET, "Texture filter set to %s")
+GS_DEFINE(GAMEPLAY_SETTINGS_FIX_TEXTURE_ISSUES,             "Fix texture issues")
+GS_DEFINE(GAMEPLAY_SETTINGS_FIX_ANIMATED_SPRITES,           "Fix sprite animations")
+GS_DEFINE(GAMEPLAY_SETTINGS_FIX_QWOP_GLITCH,                "Fix QWOP glitch")
+GS_DEFINE(GAMEPLAY_SETTINGS_FIX_ITEM_DUPLICATION_GLITCH,    "Fix item duplication glitch")
```
`Fix dupe glitch` (maybe too casual)

rr- replied:
Yeah slightly too casual IMO.

lahm86 replied:
I think this one is fine as-is 👍

lahm86 commented on src/tr1/game/game_string.def:
```diff
@@ -23,6 +15,146 @@ GS_DEFINE(OSD_TEXTURE_FILTER_SET, "Texture filter set to %s")
+GS_DEFINE(GAMEPLAY_SETTINGS_FIX_ANIMATED_SPRITES,           "Fix sprite animations")
+GS_DEFINE(GAMEPLAY_SETTINGS_FIX_QWOP_GLITCH,                "Fix QWOP glitch")
+GS_DEFINE(GAMEPLAY_SETTINGS_FIX_ITEM_DUPLICATION_GLITCH,    "Fix item duplication glitch")
+GS_DEFINE(GAMEPLAY_SETTINGS_FIX_TIHOCAN_SECRET_SOUND,       "Fix Tomb of Tihocan secret sound")
```
`Fix chain block SFX`

rr- replied:
Went with `Fix chain block sound`

lahm86 commented on src/tr1/game/game_string.def:
```diff
@@ -23,6 +15,146 @@ GS_DEFINE(OSD_TEXTURE_FILTER_SET, "Texture filter set to %s")
+GS_DEFINE(GAMEPLAY_SETTINGS_FIX_SHOTGUN_TARGETING,          "Enhanced shotgun targeting")
+GS_DEFINE(GAMEPLAY_SETTINGS_REVERT_TO_PISTOLS,              "Pistols as primary weapon")
+GS_DEFINE(GAMEPLAY_SETTINGS_MAXIMUM_SAVE_SLOTS,             "Number of save slots")
+GS_DEFINE(GAMEPLAY_SETTINGS_ENABLE_ENHANCED_SAVES,          "Save additional game information")
```
`Save effects`

rr- replied:
I wonder if it makes sense to remove this setting and force it to be always on?
Or at least, remove the GUI for it?

lahm86 replied:
Yeah, that could be a good move. AFAIK there are two cases other than the improved visual/immersion aspect it brings:
- If Lara is on fire and you save/load with it disabled, she's no longer on fire (TR2 fixed this with `g_Lara.burn`, we should probably do the same in TR1)
- If you're doing a 1HP challenge, this lets you bypass forced-damage or especially difficult areas by getting rid of embers, darts, etc on reload

rr- replied:
Oh yeah. Killing embers is a legit concern. I didn't think of it – it makes sense to keep it.

lahm86 commented on src/tr1/game/game_string.def:
```diff
@@ -23,6 +15,146 @@ GS_DEFINE(OSD_TEXTURE_FILTER_SET, "Texture filter set to %s")
+GS_DEFINE(GAMEPLAY_SETTINGS_ENABLE_SAVE_CRYSTALS,           "Enable save crystals")
+GS_DEFINE(GAMEPLAY_SETTINGS_ENABLE_WALK_TO_ITEMS,           "Animated interactions")
+GS_DEFINE(GAMEPLAY_SETTINGS_ENABLE_AUTO_ITEM_SELECTION,     "Key item pre-selection")
+GS_DEFINE(GAMEPLAY_SETTINGS_ENABLE_DEATHS_COUNTER,          "Count number of deaths")
```
`Count Lara's deaths`

lahm86 commented on src/libtrx/include/libtrx/game/game_string.def:
```diff
@@ -224,3 +200,49 @@ GS_DEFINE(CONSOLE_HELP_WIREFRAME,     "Toggles wireframe rendering.")
+GS_DEFINE(SOUND_SETTINGS_ENABLE_MUSIC_IN_MENU,                   "Main menu music")
+GS_DEFINE(SOUND_SETTINGS_ENABLE_PITCHED_SOUNDS,                  "Pitched sounds")
+GS_DEFINE(SOUND_SETTINGS_ENABLE_MUSIC_IN_INVENTORY,              "Game sounds in inventory")
+GS_DEFINE(SOUND_SETTINGS_FIX_SECRETS_KILLING_MUSIC,              "Continue music on secrets pick-ups")
```
`Layered secret music`

lahm86 commented on src/libtrx/include/libtrx/game/game_string.def:
```diff
@@ -224,3 +200,49 @@ GS_DEFINE(CONSOLE_HELP_WIREFRAME,     "Toggles wireframe rendering.")
+GS_DEFINE(SOUND_SETTINGS_ENABLE_PITCHED_SOUNDS,                  "Pitched sounds")
+GS_DEFINE(SOUND_SETTINGS_ENABLE_MUSIC_IN_INVENTORY,              "Game sounds in inventory")
+GS_DEFINE(SOUND_SETTINGS_FIX_SECRETS_KILLING_MUSIC,              "Continue music on secrets pick-ups")
+GS_DEFINE(SOUND_SETTINGS_FIX_SPEECHES_KILLING_MUSIC,             "Continue music when enemies speak")
```
`Layered enemy speech`

rr- replied:
Brilliant!

lahm86 commented on src/libtrx/include/libtrx/game/game_string.def:
```diff
@@ -224,3 +200,49 @@ GS_DEFINE(CONSOLE_HELP_WIREFRAME,     "Toggles wireframe rendering.")
+GS_DEFINE(SOUND_SETTINGS_FIX_SECRETS_KILLING_MUSIC,              "Continue music on secrets pick-ups")
+GS_DEFINE(SOUND_SETTINGS_FIX_SPEECHES_KILLING_MUSIC,             "Continue music when enemies speak")
+GS_DEFINE(SOUND_SETTINGS_PS_UZI_SFX,                             "PS1 Uzi sound")
+GS_DEFINE(SOUND_SETTINGS_UNDERWATER_MUSIC_MODE,                  "Underwater music behavior")
```
`Underwater music mode`

rr- replied:
@walkawayy suggested to split this setting to two smaller settings, one for ambient and one for normal music tracks, to save space on the too long enum values. WDYT? Not every combination is going to make sense this way but for sure it'll help with space. I wonder if there's anything more we could do here.

lahm86 replied:
I think that sounds good. They're quite involved as it stands, plus with #2870 for TR2 it'd be nice to agree on something common for inventory, underwater and perhaps save/load. Will have a think.

lahm86 replied:
Some thoughts:

- New enum: `SPECIAL_MUSIC_MODE` (or a better name)
  - `Muted`
  - `Quiet`
  - `Normal`
- `Inventory game sounds` becomes `SPECIAL_MUSIC_MODE` (the `Quiet` option here would only apply to music, not SFX)
- `Underwater music mode` becomes `SPECIAL_MUSIC_MODE`
- `Load music track` becomes a `bool`; ambient and regular tracks treated the same
  - An aside, perhaps we should rename this to `Restore music on load`

Then have one additional `SPECIAL_MUSIC_MODE` setting for ambient as a whole, and have it target both inventory and underwater. Potential names:
- `Ambient music dimming`
- `Situational ambient mode`
- `Dynamic ambient mode`

I think then the only thing we'd lose is the specific handling of ambient when loading a save. Maybe that's OK? IMHO, the difference between restoring an ambient timestamp and not, is not all that noticeable; it's much clearer with a regular track what's happening.

rr- replied:
My personal preference would be to not restore the ambient tracks. Agreed about `Restore music on load`.
`Inventory game sounds` set to `quiet` I imagine would mute the music to half the volume so it's a bit confusing.

I feel `SPECIAL_MUSIC_MODE` may be a bit confusing. WDYT about this?
Enums:
```c
typedef {
    AUDIO_NONE    = 0b0000,
    AUDIO_AMBIENT = 0b0001,
    AUDIO_MUSIC   = 0b0010,
    AUDIO_SOUND   = 0b0100,
} AUDIO_SOURCES;
```
as a bit flag,

```c
typedef {
    AUDIO_MUTED,
    AUDIO_QUIET,
    AUDIO_FULL_VOLUME,
} AUDIO_VOLUME;
```
(I hate the name `AUDIO` as it clashes with libtrx's game-agnostic module and this is very much game-related, but the idea is to let it refer to both music and sounds.)

The settings:
Inventory audio: Muted / Music only / Play all (with a custom setter that sets `AUDIO_SOURCES`)
Inventory volume: Muted / Quiet / Normal (linked to `AUDIO_VOLUME`)
Underwater music: None / Music only / All (with a custom setter that sets `AUDIO_SOURCES`)
Underwater volume: Muted / Quiet / Normal (linked to `AUDIO_VOLUME`)

There's a kind of duplication between "Play none" and "Play muted" but maybe it can be tweaked, just an idea.

lahm86 replied:
This sounds good to me. How about using `acoustic` in place of `audio` for the enums?

rr- replied:
Hmm. Maybe… `PLAYBACK`? Or `AUDIO_TARGET`.
Also with my proposal it will be impossible to have the sounds play normally underwater and music at half volume, without doing some assumptions (eg `underwater_playback: PLAYBACK_TARGET_ALL`, `underwater_volume: PLAYBACK_VOLUME_MUTED`, and yet we expect only the music to play at half volume while the sounds at full volume). I think that's somewhat acceptable.

lahm86 replied:
`PLAYBACK` sounds good. And agreed, the second point is acceptable IMO.

lahm86 commented on src/libtrx/include/libtrx/game/game_string.def:
```diff
@@ -224,3 +200,49 @@ GS_DEFINE(CONSOLE_HELP_WIREFRAME,     "Toggles wireframe rendering.")
+GS_DEFINE(SOUND_SETTINGS_SOUND_VOLUME,                           "\\{icon music} Music")
+GS_DEFINE(SOUND_SETTINGS_MUSIC_VOLUME,                           "\\{icon sound} Sound")
+GS_DEFINE(SOUND_SETTINGS_MUSIC_LOAD_CONDITION,                   "Load music track")
+GS_DEFINE(SOUND_SETTINGS_LOAD_MUSIC_TRIGGERS,                    "Remember played music")
```
I feel this could maybe be combined with `Fix floor data issues`, but that's maybe controversial. It's a little confusing alongside `Load music track`. Perhaps just `Remember triggers` would do with it being in the sound dialog, and later a description could provide more detail?

rr- replied:
I think it's one of these settings that make sense to be always on. WDYT about removing it from the GUI?

rr- replied:
Or maybe move it to gameplay fixes "Fix one-shot music triggers"

lahm86 replied:
Moving it to gameplay fixes sounds good - more relevant there and it removes any confusion with loading the music track.

rr- replied:
Moved

lahm86 commented on src/tr1/game/game_string.def:
```diff
@@ -23,6 +15,146 @@ GS_DEFINE(OSD_TEXTURE_FILTER_SET, "Texture filter set to %s")
+GS_DEFINE(GAMEPLAY_SETTINGS_REVERT_TO_PISTOLS,              "Pistols as primary weapon")
+GS_DEFINE(GAMEPLAY_SETTINGS_MAXIMUM_SAVE_SLOTS,             "Number of save slots")
+GS_DEFINE(GAMEPLAY_SETTINGS_ENABLE_ENHANCED_SAVES,          "Save additional game information")
+GS_DEFINE(GAMEPLAY_SETTINGS_CHANGE_PIERRE_SPAWN,            "Change Pierre spawn behavior")
```
`Change Pierre spawn mode`

lahm86 reviewed (commented):
Added some potential text alternatives for some of the longer ones (although they're not great admittedly, but see what you think). Perhaps we can get away with a snappier/casual tone and once there is the availability of descriptions, it'll be enough for clarity.

lahm86 commented:
A minor UI issue I noticed - in the graphics dialog for example, the first two tabs have the bottom arrow visible for scrolling, but because the third tab doesn't have scroll, its height is slightly less than the others. So you get a minor dialog shift when selecting this tab.

rr- commented:
> A minor UI issue I noticed - in the graphics dialog for example, the first two tabs have the bottom arrow visible for scrolling, but because the third tab doesn't have scroll, its height is slightly less than the others. So you get a minor dialog shift when selecting this tab.

Thanks – fixed!

rr- committed:
config: replace underwater music mode with volumes

rr- commented:
Replaced underwater music mode setting with volumes. Removed this setting from the config tool both for TR1 and TR2; my plan is to follow up after merging this PR with a similar dialog updates for TR2 to justify the removal.

To do:
- [x] Convert sound and music volume settings to show percentages
- [x] Pause music in inventory
- [x] Inventory volumes

rr- committed:
config: make sound and music volume percentages

rr- committed:
ui/sound: flip pause music in inventory option

rr- committed:
ui/sound: add inventory music volume control

rr- commented:
@aredfan @lahm86 updated. I think it looks neat. I'm open for more suggestions on reordering the items, moving them around or changing their text. We now have an easy way to flip the booleans inside the settings dialog UI thanks to `COT_INVERTED_BOOL`.

aredfan commented:
A couple of issues I've come across:

- [x] The music/sound volumes can go below 0% for both games. In the case of TR2X it displays `-10%`.
- [x] If it's relevant, I noticed that the PDA doesn't animate in the inventory for the TR1/UB levels or in the gym.

lahm86 commented:
Looking great!

WDYT about more consistency around bool option text? For example we have ones like `Skyboxes`, `Breeze`, etc and we have others like `Enable cutscenes`, `Enable buffering` etc. Should we drop `Enable`, given the `On`/`Off` text makes it quite clear?

Below is a proposal for reordering some items, loosely based on grouping reasonably related things (my take on it at least, feel free to adjust 😄).

<details><summary>Sound</summary>

- Music volume
- Sound volume
- Main menu music
- Pitched sounds
- Restore music on load
- Layered secret music
- Layered enemy speech
- PS1 Uzi sound
- Pause sound in inventory
- Inventory ambient volume
- Inventory music volume
- Underwater ambient volume
- Underwater music volume

</details>

<details><summary>Graphics</summary>

**UI**
Drop the `UI` from the start of `Text scale`, `Bar scale` and `Scroll wrap`

**Rendering**
- FPS
- Anisotropy filter
- Texture filter
- Trapezoid filter
- Reflections
- FBO filter
- Resolution
- Render mode
- VSync
- Brightness
- Screenshot format

</details>

<details><summary>Gameplay</summary>

**General**
- Enable legal
- Enable FMVs
- Enable demo mode
- Enable game modes
- Enable loading screens
- Enable cutscenes
- Enable save crystals
- Number of save slots
- Save effects
- Key item pre-selection
- Item examination
- Timer counts in inventory
- Count Lara's deaths
- Stat detail mode _(perhaps rename `Stat` to `Statistics`)_
- Level statistics in compass
- Finals statistics screen
- Remove dead-T-Rex collision
- Change Pierre spawn mode
- Restore PS1 enemies
- Pistols as primary weapon
- Enhanced shotgun targeting

**Controls**
- Responsive passport
- Hotkey item use
- Animated interactions _(moved from General)_
- Enhanced sidesteps
- Look while running
- Responsive jumping
- Jump twists
- Lean jumping
- Swing cancels
- Underwater roll
- Smooth swimming
- Response swim cancel
- Wading
- Weapon lock mode
- Target change
- Inverted look
- Camera speed
- Enable buffering

**Fixes**
- Fix floor data issues
- Fix texture issues
- Fix item rotation issues
- Fix one-shot music triggers
- Fix sprite animations
- Fix alligator AI
- Fix bear AI
- Fix chain block sound
- Fix breakable floor falls
- Fix bridge collision
- Fix QWOP glitch
- Fix item duplication glitch
- Fix wall jump glitch

</details>

One final thought with the inverted bool now available - WDYT about `Pistols as primary weapon` becoming `Remember guns between levels` (or similar wording)? So the emphasis is on carrying over guns rather than reverting to pistols.

rr- committed:
tr1/ui: reorder settings

rr- commented:
Fixed the PDA model issue and reordered most of the items, although with small deviations from Lahm's suggestion (mostly tried to push very OG-specific things such as PS1 uzis, bear AI bottom-most).
I'm not sure how to consistently reproduce the -10% issue aredfan found – I only once got the TR2 UI to show -0% as a rounding error.

aredfan commented:
> Fixed the PDA model issue

Thank you. The PDA looks good. 👍 

> I'm not sure how to consistently reproduce the -10% issue aredfan found – I only once got the TR2 UI to show -0% as a rounding error.

I think the game need be started with the sound/music volume in the 10-60% range, in my case it's 50%. I'm unsure if there's any other steps needed to replicate the issue.

For TR2X, if the sound/music volume is showing -10%, these are the values in the TR2X.json5 file:

```
  "sound_volume": -0.1,
  "music_volume": -0.1,
```

<details><summary>Click here to see the video</summary>

https://github.com/user-attachments/assets/7ba41a42-3934-42cd-9bb1-8d96b8dd86a9

</details>

For TR1X, the sound/music volume remains at 0% which is a good sign, I wrongly assumed the value can go lower.

```
  "music_volume": 0.0,
  "sound_volume": 0.0,
```


<details><summary>Click here to see the video</summary>

https://github.com/user-attachments/assets/e0eb7ef0-0a7b-4b19-a17e-776a0ec315c5

</details>

I've also come across a second bug, if I delete the `TR1X.json5` or `TR2X.json5` files, the game doesn't start and instead creates a `.dmp` file in the directory.

rr- committed:
ui/settings: fix overflows in handling float values

rr- commented:
Thanks @aredfan fixed the crash, and I think I also fixed the percentage ranges problem. (It could've introduced regressions with other decimal values.)

aredfan commented:
> Thanks @aredfan fixed the crash, and I think I also fixed the percentage ranges problem. (It could've introduced regressions with other decimal values.)

Anytime, and thank you. 👍 

I can confirm the percent range described above is fixed. There is a second issue which I think you've encountered too, it only affects TR2X. If I start up the game with the volume in the 70-100% range, then reduce said volume all the way down, it reaches `-0%`.

This is how the values appear in the TR2X.json5.

```
  "sound_volume": -0.0,
  "music_volume": -0.0,
```

_______________

Btw I noticed that the PDA is present within the inventory of the demo (`--demo-pc`) and level (`-l`) but doesn't animate in either.

______________

**Edit:** I forgot to mention that the crash is fixed too.

rr- commented:
@aredfan fixed both issues :relieved:

aredfan reviewed (approved):
Everything LGTM.

Thank you @rr-  👍

lahm86 reviewed (approved):
Awesome 😌
